server/storage/players.lua

```diff
@@ -382,6 +410,124 @@ local function cleanPlayerGroups()
+        lib.print.error(('Tried to update player data field %s when it doesn\'t exist. Value: %s'):format(hasSubKeys and key[1] or key, value))
+        return
+    end
+
+    if hasSubKeys and not jsonPlayerFields[key[1]] then
+        lib.print.error(('Tried to update player data field %s as a json object when it isn\'t one'):format(key[1]))
+        return
+    end
+
+    value = type(value) == 'table' and json.encode(value) or value
+
+    local currentTable = isUpdating and playerDataUpdateQueue or collectedPlayerData
+    if not currentTable[citizenid] then
+        currentTable[citizenid] = {}
+    end
+
+    currentTable[citizenid][hasSubKeys and key[1] or key] = hasSubKeys and {} or value
+
+    if hasSubKeys then
+        local current = currentTable[citizenid][key[1]]
+        if #key > 2 then
+            -- We don't check the last one because otherwise we lose the table reference
+            for i = 2, #key - 1 do
+                if not current[key[i]] then
+                    current[key[i]] = {}
+                end
+
+                current = current[key[i]]
+            end
+        end
+
+        current[key[#key]] = value
+    end
+end
+
+---@param key string
+---@param nestedTable table<string, any>
+---@param path string?
+---@param citizenid string
+local function updateNestedPlayerData(key, nestedTable, citizenid, path)
+    for k, v in pairs(nestedTable) do
+        local currentPath = path and ('%s.%s'):format(path, k) or k
+        if type(v) == 'table' then
+            updateNestedPlayerData(key, v, citizenid, currentPath)
+        else
+            local query = ('UPDATE players SET %s = JSON_SET(%s, "$.%s", ?) WHERE citizenid = ?'):format(key, key, currentPath)
+            MySQL.prepare.await(query, { v, citizenid })
+        end
+    end
+end
+
+local function sendPlayerDataUpdates()
+    if isUpdating then return end
+
+    -- We wait on a single player to be updated to not mess with the collectedPlayerData table whilst it's updating
+    while isPlayerUpdating do
+        Wait(10)
+    end
+
+    -- We implement this to ensure when updating no values are added to our updating sequence to prevent data loss by accidentally skipping over it
+    isUpdating = true
+
+    for citizenid, playerData in pairs(collectedPlayerData) do
+        for key, data in pairs(playerData) do
+            if type(data) == 'table' then
+                updateNestedPlayerData(key, data, citizenid)
+            else
+                local query = ('UPDATE players SET %s = ? WHERE citizenid = ?'):format(key)
+                MySQL.prepare.await(query, { data, citizenid })
+            end
+        end
+    end
+
+    collectedPlayerData = playerDataUpdateQueue
+    playerDataUpdateQueue = {}
+    isUpdating = false
+end
+
+---@param citizenid string
+local function forcePlayerDataUpdate(citizenid)
+    -- We don't need to update a single player when everyone is already getting an update
+    if isUpdating then return end
```

Comment:
If this isn't self explanatory then some refactoring is necessary.

Reply:
You underestimate how much people can understand code and who can view this code.